[XLA:GPU][AutoSharding] Replace hashmap with btree map to determine the iteration order

lausannel opened the pull request:
This PR addresses one of the problems mentioned in #7248, regarding the non-deterministic iteration order of `StableHashMap`.

To resolve this, this PR proposes the following changes:

- Replaced `::absl::flat_hash_map` with `::absl::btree_map` within the `StableHashMap` implementation.
- Added new tests to validate the deterministic iteration order of the `StableHashMap`.

dimitar-asenov commented on xla/hlo/experimental/auto_sharding/auto_sharding_strategy.h:
```diff
@@ -42,7 +43,7 @@ constexpr double kInfinityCost = 1e20;
 // Type alias
 template <typename Key, typename Value>
-using StableHashMap = ::absl::flat_hash_map<Key, Value>;
+using StableHashMap = ::absl::btree_map<Key, Value>;
```
Please add a test that checks that `StableHashMap` now indeed has deterministic order. E..g a test where you insert 10 pairs in a random order and expect to read them in a known order.

cheshire replied:
TBH I think this is fine, we never had flat_hash_map here before (it's different internally), and not everything is deterministic yet, so this fix clearly makes it incrementally better.

dimitar-asenov replied:
I'm not opposed to the solution, just want to have a test that clearly states that whatever `StableHashMap` is defined to be, it should be deterministic.

dimitar-asenov commented:
The description says that this fixes https://github.com/openxla/xla/issues/7248, but looking at that issue, it seems that there is more that needs to be done to fix it. Can you please explain if this PR really fixes the issue? If not, then please adjust the description.

cheshire commented:
> The description says that this fixes https://github.com/openxla/xla/issues/7248,

Agreed with the comment, would not mention the bug as "fixed".

lausannel commented:
@cheshire @dimitar-asenov Sorry for the lack of clarity. Thank you for your review comments. I have updated the PR description and added the corresponding tests. Could you take another look?
